modules/raft/src/main/java/org/apache/ignite/raft/jraft/NodeManager.java

```diff
@@ -14,24 +14,158 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
 package org.apache.ignite.raft.jraft;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiPredicate;
 import java.util.stream.Collectors;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.network.ClusterService;
+import org.apache.ignite.raft.jraft.core.Scheduler;
 import org.apache.ignite.raft.jraft.entity.NodeId;
 import org.apache.ignite.raft.jraft.entity.PeerId;
+import org.apache.ignite.raft.jraft.option.NodeOptions;
+import org.apache.ignite.raft.jraft.rpc.CoalescedHeartbeatRequestBuilder;
+import org.apache.ignite.raft.jraft.rpc.Message;
+import org.apache.ignite.raft.jraft.rpc.RpcClient;
+import org.apache.ignite.raft.jraft.rpc.RpcRequests.AppendEntriesRequest;
+import org.apache.ignite.raft.jraft.rpc.RpcRequests.CoalescedHeartbeatResponse;
+import org.apache.ignite.raft.jraft.rpc.impl.IgniteRpcClient;
 import org.apache.ignite.raft.jraft.util.OnlyForTest;
 
 /**
  * Raft nodes manager.
  */
-public class NodeManager {
+public class NodeManager implements Lifecycle<NodeOptions> {
+    private static final IgniteLogger LOG = Loggers.forClass(NodeManager.class);
+
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
     private final ConcurrentMap<NodeId, Node> nodeMap = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, List<Node>> groupMap = new ConcurrentHashMap<>();
+    private final ConcurrentMap<PeerId, Queue<Object[]>> coalesced = new ConcurrentHashMap<>();
+
+    /** Node options. */
+    private NodeOptions options;
+    /** Task scheduler. */
+    private Scheduler scheduler;
+    /** Rpc client. */
+    private final RpcClient rpcClient;
+    /** Message factory. */
+    private RaftMessagesFactory messagesFactory;
+    /** Predicate to block a heartbeat messages. */
+    private BiPredicate<Message, PeerId> blockPred;
+
+    public NodeManager(ClusterService service) {
+        rpcClient = new IgniteRpcClient(service);
+    }
+
+    @Override
+    public boolean init(NodeOptions opts) {
+        options = opts;
+        scheduler = opts.getScheduler();
+        messagesFactory = opts.getRaftMessagesFactory();
+
+        scheduler.schedule(this::onSentHeartbeat , opts.getElectionTimeoutMs(), TimeUnit.MILLISECONDS);
```

Comment:
This scheme is not efficient.
Currently each raft group generate it's own timer event in r.startHeartbeatTimer(Utils.nowMs());
Which means events multiplication linear to groups number and excessive CPU usage.
Instead, a single event per all groups can be triggered.

Reply:
Probably we can create another ticket for this? Or you insist to do this improvement in this one.

Reply:
I'm ok with another ticket.

Reply:
https://issues.apache.org/jira/browse/IGNITE-24789